# Importing Libraries and Setup

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# creating path variable
path = r'/Users/paigepearson/Desktop/Data Analytics/Immersion Course/Achievement 4/Instacart Basket Analysis'

In [5]:
# importing products_checked.csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'), index_col = 0)

## Step 3 - In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

In [7]:
# importing orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_combined.pkl'))

## Step 4 - Check the shape of the imported dataframe

In [9]:
df_ords_prods_combined.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2539329        1    prior             1                  2   
2   2539329        1    prior             1                  2   
3   2539329        1    prior             1                  2   
4   2539329        1    prior             1                  2   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  8                    NaN       14084                  2   
2                  8                    NaN       12427                  3   
3                  8                    NaN       26088                  4   
4                  8                    NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [10]:
df_ords_prods_combined.shape

(32434489, 11)

In [13]:
df_prods.shape

(49672, 5)

Note: In part 1 of this task the df_ords_merged had 32,434,489 rows and 11 columns. df_ords_prods_combined has the same number of rows and columns and are therefore the same.

## Step 5 - Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [12]:
# checking the dataframes to determine how to merge the two
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Note: product_id is a column that can be used as a key. _merge column may be an issue from previous work, dropping this as it is not needed going forward and may hamper the new merge.

In [14]:
# dropping _merge column leftover from previous merge
df_ords_prods_combined = df_ords_prods_combined.drop(columns = ['_merge'])

In [15]:
# checking work
df_ords_prods_combined.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2539329        1    prior             1                  2   
2   2539329        1    prior             1                  2   
3   2539329        1    prior             1                  2   
4   2539329        1    prior             1                  2   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  8                    NaN       14084                  2   
2                  8                    NaN       12427                  3   
3                  8                    NaN       26088                  4   
4                  8                    NaN       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [16]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [17]:
# merging df_prods and df_ords_prods_combined 
df_prods_merged = df_ords_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

In [18]:
# checking work
df_prods_merged.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [19]:
df_prods_merged.shape

(32404859, 15)

## Step 6 - Confirm the results of the merge using the merge flag.

In [20]:
# checking merge results, told in lesson that this merge would be a full merge and there shouldn't be any left or right only merges
df_prods_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

## Step 7 - Export the newly created dataframe as orders_products_merged in a suitable format (taking into consideration the size).

In [21]:
# Exporting with pickle as the size may kill my computer otherwise
df_prods_merged.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged.pkl'))